### PR TITLE
[PWX-32202] storkctl perform failover CLI changes

### PR DESCRIPTION
**Pull request by olavangad-px**
**What type of PR is this?**
>feature
>unit-test

**What this PR does / why we need it**:
To add storkctl perform failover cli command. 
This command creates an action CR with actionType:failover for which we will add logic in the ActionController to perform the actual failover.

Another change is that we will create static copy of migrationSchedules in remote cluster irrespective of autoSuspend value

**Does this PR change a user-facing CRD or CLI?**:
Yes, it adds a new storkctl command `storkctl perform failover`
```
[root@olavangad-84-0 ~]# storkctl perform failover -h
Initiate failover of the given migration schedule

Usage:
  storkctl perform failover [flags]

Flags:
      --exclude-namespaces strings   Specify the comma-separated list of subset namespaces to be skipped during the failover. By default, all namespaces part of the migrationSchedule are failed over
  -h, --help                         help for failover
      --include-namespaces strings   Specify the comma-separated list of subset namespaces to be failed over. By default, all namespaces part of the migrationSchedule are failed over
  -m, --migration-reference string   Specify the migration schedule to failover. Also specify the namespace of this migrationSchedule using the -n flag
      --skip-deactivate-source       If present, applications in the source cluster will not be scaled down as part of the failover.

Global Flags:
      --alsologtostderr                   log to standard error as well as files
      --burst int                         restrict number of k8s API requests from stork (default 2000)
      --context string                    name of the kubeconfig context to use
      --kubeconfig string                 path to the kubeconfig file to use for CLI requests
      --log_backtrace_at traceLocations   when logging hits line file:N, emit a stack trace
      --log_dir string                    If non-empty, write log files in this directory
      --log_link string                   If non-empty, add symbolic links in this directory to the log files
      --logbuflevel int                   Buffer log messages logged at this level or lower (-1 means don't buffer; 0 means buffer INFO only; ...). Has limited applicability on non-prod platforms.
      --logtostderr                       log to standard error instead of files
  -n, --namespace string                  If present, the namespace scope for this CLI request (default "default")
  -o, --output string                     output format. One of: table|json|yaml (default "table")
      --qps int                           restrict number of k8s API requests from stork (default 1000)
      --stderrthreshold severityFlag      logs at or above this threshold go to stderr (default 2)
  -v, --v Level                           log level for V logs
      --vmodule vModuleFlag               comma-separated list of pattern=N settings for file-filtered logging
  -w, --watch                             watch stork resources

```

**Is a release note needed?**:
Yes, TBD

**Does this change need to be cherry-picked to a release branch?**:
Yes, 24.2.0

**Testing Details**:
1. Have tested that remote copy of migrationSchedule is created irrespective of autoSuspend value.
2. Have added comprehensive unit tests for storkctl perform failover and tested the corner cases manually as well.

### CODE REVIEW

File: pkg/migration/controllers/migrationschedule.go [hunk 1]
```diff
@@ -149,50 +149,61 @@ func (m *MigrationScheduleController) handle(ctx context.Context, migrationSched
...
+			remoteMigrSched.ResourceVersion = ""
+			remoteMigrSched.UID = ""
+			if remoteMigrSched.Annotations == nil {
+				remoteMigrSched.Annotations = make(map[string]string)
+			}
+			remoteMigrSched.Annotations[StorkMigrationScheduleCopied] = "true"
+			suspend := true
+			remoteMigrSched.Spec.Suspend = &suspend
+			remoteMigrSched.Status = stork_api.MigrationScheduleStatus{}
+			if _, err := remoteOps.CreateMigrationSchedule(remoteMigrSched); err != nil {
```
Comment (diptiranjanpx) [inline pkg/migration/controllers/migrationschedule.go:177]: Only return of error is not  AlreadyExists.
Also use the migration schedule copy returned by the server from there onwards instead of `remoteMigrSched`.

File: pkg/migration/controllers/migrationschedule.go [hunk 2]
```diff
@@ -149,50 +149,61 @@ func (m *MigrationScheduleController) handle(ctx context.Context, migrationSched
...
+			suspend := true
+			migrationSchedule.Spec.Suspend = &suspend
+			msg := "Suspending migration schedule since migrated apps on remote cluster are active"
+			m.recorder.Event(migrationSchedule,
+				v1.EventTypeWarning,
+				"Suspended",
+				msg)
+			log.MigrationScheduleLog(migrationSchedule).Warn(msg)
+			// deactivate apps in namespace
+			// TODO: suspend deploy/sts,crds
```
Comment (diptiranjanpx) [inline pkg/migration/controllers/migrationschedule.go:205]: nit: Let's remove this line.

File: pkg/storkctl/action.go [hunk 3]
```diff
@@ -20,99 +20,144 @@ const (
...
-	actionList, err := storkops.Instance().ListActions(namespace)
-	if err != nil {
-		util.CheckErr(err)
-		return nil
+// given the name, namespace, actionType and actionParameters create an Action CR
+func createActionCR(actionName string, namespace string, actionType storkv1.ActionType, actionParameters storkv1.ActionParameter) (*storkv1.Action, error) {
+	action := storkv1.Action{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      actionName,
+			Namespace: namespace,
```
Comment (diptiranjanpx) [inline pkg/storkctl/action.go:142]: Should we put an annotation "created-by: stork" to figure out if it has been created by storkctl ?

File: pkg/storkctl/action.go [hunk 4]
```diff
@@ -20,99 +20,144 @@ const (
...
-	return strings.Join([]string{string(action), time.Now().Format(nameTimeSuffixFormat)}, "-")
+func getActionStatusMessage(action *storkv1.Action) string {
+	return fmt.Sprintf("To check %v status use the command : `storkctl get %v %v -n %v`", action.Spec.ActionType, action.Spec.ActionType, action.Name, action.Namespace)
 }
 
-func getDescribeActionMessage(action *storkv1.Action) string {
-	return "To check Action status use: " +
-		fmt.Sprintf("kubectl describe action %v -n %v", action.Name, action.Namespace)
+func getActionName(actionType storkv1.ActionType, referenceResourceName string) string {
+	return strings.Join([]string{string(actionType), referenceResourceName, GetCurrentTime().Format(nameTimeSuffixFormat)}, "-")
```
Comment (diptiranjanpx) [inline pkg/storkctl/action.go:162]: let's trim <actionType>-<referenceResourceName> so that the whole string including the time fits into the max chars. Also let's put these kind of functions in utils.

File: pkg/migration/controllers/migrationschedule.go (hunk 2 above)
Comment (olavangad-px) [reply pkg/migration/controllers/migrationschedule.go:205]: You mean the TODO line?

File: pkg/migration/controllers/migrationschedule.go (hunk 1 above)
Comment (olavangad-px) [reply pkg/migration/controllers/migrationschedule.go:177]: This particular branch will be entered only if remote migration schedule doesn't already exist. 
So err caught here cannot be AlreadyExists.

2nd point makes sense. Will address.

File: pkg/storkctl/action.go (hunk 3 above)
Comment (olavangad-px) [reply pkg/storkctl/action.go:142]: Since we are not adding such an annotation in other migration objects clusterpair, migrationschedule 
I think we can skip this for now.
If we decide in the future we can make the change across all objects.

File: pkg/storkctl/action.go (hunk 4 above)
Comment (olavangad-px) [reply pkg/storkctl/action.go:162]: I will trim it according to 253 chars

**Review by diptiranjanpx — APPROVED** [review]

File: pkg/storkctl/action_test.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+package storkctl
+
+import (
+	"fmt"
+	storkops "github.com/portworx/sched-ops/k8s/stork"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"testing"
+	"time"
```
Comment (pp511) [inline pkg/storkctl/action_test.go:12]: nit..Let's group two groups here too.

File: pkg/migration/controllers/migrationschedule.go [hunk 6]
```diff
@@ -149,50 +149,58 @@ func (m *MigrationScheduleController) handle(ctx context.Context, migrationSched
...
-				remoteMigrSched.Annotations[StorkMigrationScheduleCopied] = "true"
-				suspend := true
-				remoteMigrSched.Spec.Suspend = &suspend
-				remoteMigrSched.Status = stork_api.MigrationScheduleStatus{}
-				if _, err := remoteOps.CreateMigrationSchedule(remoteMigrSched); err != nil {
-					return err
-				}
 
-			} else if err != nil {
+		// create migrationSchedule in remote cluster irrespective of autoSuspend value
```
Comment (pp511) [inline pkg/migration/controllers/migrationschedule.go:153]: I am missing the point. Help me understand why we need to copy the migrationschedules?

**Review by pp511 — APPROVED** [review]
Looks good. Just one question and some nit comments.

File: pkg/storkctl/action.go [hunk 7]
```diff
@@ -2,6 +2,7 @@ package storkctl
 
 import (
 	"fmt"
+	"k8s.io/apimachinery/pkg/util/validation"
```
[the remark this answers is not in the document]
Comment (olavangad-px) [reply pkg/storkctl/action.go:5]: will address

File: pkg/storkctl/action.go [hunk 8]
```diff
@@ -20,99 +21,150 @@ const (
...
+			if len(referenceMigrationSchedule) == 0 {
+				util.CheckErr(fmt.Errorf("referenceMigrationSchedule name needs to be provided for failover"))
+				return
+			}
+			// namespace of the migrationSchedule is provided by user using the -n / --namespace global flag
+			migrationScheduleNs := cmdFactory.GetNamespace()
+			migrSchedObj, err := storkops.Instance().GetMigrationSchedule(referenceMigrationSchedule, migrationScheduleNs)
 			if err != nil {
-				util.CheckErr(err)
+				util.CheckErr(fmt.Errorf("unable to find the referenceMigrationSchedule %v in the %v namespace", referenceMigrationSchedule, migrationScheduleNs))
```
[the remark this answers is not in the document]
Comment (olavangad-px) [reply pkg/storkctl/action.go:69]: makes sense. will address

File: pkg/storkctl/action.go [hunk 9]
```diff
@@ -20,99 +21,150 @@ const (
...
+			// at most one of exclude-namespaces or include-namespaces can be provided
+			if len(includeNamespaceList) != 0 && len(excludeNamespaceList) != 0 {
+				util.CheckErr(fmt.Errorf("can provide only one of --include-namespaces or --exclude-namespaces values at once"))
+				return
+			} else if len(includeNamespaceList) != 0 {
+				if isSubset, nonSubsetStrings := isSubset(includeNamespaceList, migrationNamespaces); isSubset {
+					// Branch 1: Only failover some of the namespaces being migrated by the given migrationSchedule
+					namespaceList = includeNamespaceList
+				} else {
+					util.CheckErr(fmt.Errorf("provided namespaces %v are not a subset of the namespaces being migrated by the given migrationSchedule", nonSubsetStrings))
```
[the remark this answers is not in the document]
Comment (olavangad-px) [reply pkg/storkctl/action.go:97]: okay will address

File: pkg/storkctl/action.go [hunk 10]
```diff
@@ -20,99 +21,150 @@ const (
...
+			printMsg(getActionStatusMessage(action), ioStreams.Out)
 		},
 	}
-	return failoverCommand
-}
-
-func isActionIncomplete(action *storkv1.Action) bool {
-	return action.Status.Status == storkv1.ActionStatusScheduled || action.Status.Status == storkv1.ActionStatusInProgress
+	performFailoverCommand.Flags().BoolVar(&skipDeactivateSource, "skip-deactivate-source", false, "If present, applications in the source cluster will not be scaled down as part of the failover.")
+	performFailoverCommand.Flags().StringVarP(&referenceMigrationSchedule, "migration-reference", "m", "", "Specify the migration schedule to failover. Also specify the namespace of this migrationSchedule using the -n flag")
```
[the remark this answers is not in the document]
Comment (olavangad-px) [reply pkg/storkctl/action.go:132]: yes, will go through the entire PR once again and make sure that i use either one of migration schedule or Migration Schedule everywhere

File: pkg/storkctl/action_test.go (hunk 5 above)
Comment (olavangad-px) [reply pkg/storkctl/action_test.go:12]: will review in all the files and make sure this is taken care of

File: pkg/migration/controllers/migrationschedule.go (hunk 6 above)
Comment (olavangad-px) [reply pkg/migration/controllers/migrationschedule.go:153]: perform failover command is run in the context of destination cluster with the flag --migration-reference <migrationScheduleName>
And the migrationSchedule is then used to get the clusterPairName and create last mile migration object. 
Obviously for this to work, the reference migrationschedule should exist in the destination cluster.

As of stork 23.11, migrationschedule's staticCopy is created in the remote cluster only in case the original source migrationSchedule has autoSuspend set to true.
From stork 24.2 onwards, we will always create the staticCopy in the remote cluster.

File: pkg/storkctl/action.go [hunk 11]
```diff
@@ -20,99 +21,150 @@ const (
...
+					return
 				}
 			}
-			if failedToStart {
+			migrationNamespaceList := migrSchedObj.Spec.Template.Spec.Namespaces
+			migrationNamespaceSelectors := migrSchedObj.Spec.Template.Spec.NamespaceSelectors
+			// update the migrationNamespaces list by fetching namespaces based on provided label selectors
+			migrationNamespaces, err := getMigrationNamespaces(migrationNamespaceList, migrationNamespaceSelectors)
+			if err != nil {
+				util.CheckErr(fmt.Errorf("unable to get the namespaces based on the --namespace-selectors in the provided migrationSchedule: %v", err))
```
[the remark this answers is not in the document]
Comment (olavangad-px) [reply pkg/storkctl/action.go:85]: The helper function being called here, `getMigrationNamespaces` returns the migrationNamespaceList itself in case selectors is empty. So no possibility of an error in that scenario.
The function's responsibility is to append the namespaces based on the selectors to the provided list of namespaces. So errors can only arise if we are unable to get the namespaces based on the selectors.
```
namespaces, err := core.Instance().ListNamespaces(map[string]string{key: val})
```
 That's why the error message conveys the same.
